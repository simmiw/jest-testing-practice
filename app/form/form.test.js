import Form from "./page";
import {
  render,
  fireEvent,
  screen,
  getByText,
  waitFor,
} from "@testing-library/react";

jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: jest.fn(),
    refresh: jest.fn(),
  }),
}));

global.fetch = jest.fn(() =>
  Promise.resolve({
    status: 201,
  })
);

describe("form component", () => {
  it("should render the form", async () => {
    const sut = render(<Form />);
    expect(sut.container).toMatchSnapshot();
  });

  it("should display the validation error message on invalid input", async () => {
    render(<Form />);
    fireEvent.click(screen.getByText("Submit"));
    expect(screen.getByText("Name is required")).toBeVisible();
    expect(screen.getByText("Age is required")).toBeVisible();
    expect(screen.getByText("Country is required")).toBeVisible();

    fireEvent.change(screen.getByLabelText(/Name/), {
      target: { value: "123" },
    });
    fireEvent.change(screen.getByLabelText(/Age/), {
      target: { value: "-1" },
    });
    fireEvent.change(screen.getByLabelText(/Country/), {
      target: { value: "123" },
    });

    fireEvent.click(screen.getByText("Submit"));
    await waitFor(() => {
      expect(      
        screen.getByText("Name should only contain letters")
      ).toBeVisible();
      expect(screen.getByText("Age should be a valid number")).toBeVisible();
      expect(
        screen.getByText("Country should only contain letters")
      ).toBeVisible();
    });
  });

  it("should submit the form on valid input", async () => {
    render(<Form />);

    fireEvent.change(screen.getByLabelText(/Name/), {
      target: { value: "Simmi" },
    });
    fireEvent.change(screen.getByLabelText(/Age/), { target: { value: "28" } });
    fireEvent.change(screen.getByLabelText(/Country/), {
      target: { value: "India" },
    });

    fireEvent.click(screen.getByText("Submit"));

    await waitFor(() => {
      expect(screen.getByText("Submit")).toBeDisabled();
    });
    expect(fetch).toHaveBeenCalled();
  });
});
