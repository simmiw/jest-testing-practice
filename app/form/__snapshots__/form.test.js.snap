// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form component should render the form 1`] = `
<div>
  <div
    class="flex flex-wrap justify-center my-10"
  >
    <form
      class="bg-white shadow-md rounded px-8 pt-6 pb-8 mb-4 w-full max-w-sm"
      novalidate=""
    >
      <label
        class="text-gray-700 text-sm font-bold mb-2"
      >
        <span>
          Name:
        </span>
        <input
          class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline 
              "
          name="name"
          placeholder="Enter Name"
          required=""
          type="text"
        />
      </label>
      <label
        class="block text-gray-700 text-sm font-bold mb-2"
      >
        <span>
          Age:
        </span>
        <input
          class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline "
          name="age"
          placeholder="Enter Age"
          required=""
          type="number"
        />
      </label>
      <label
        class="block text-gray-700 text-sm font-bold mb-2"
      >
        <span>
          Country:
        </span>
        <input
          class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline "
          name="country"
          placeholder="Enter Country"
          required=""
          type="text"
        />
      </label>
      <button
        class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
        name="isLoading"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
